gobblin-yarn/src/main/java/org/apache/gobblin/yarn/YarnService.java

```diff
@@ -229,6 +248,21 @@ public void handleContainerShutdownRequest(ContainerShutdownRequest containerShu
     }
   }
 
+  /**
+   * Request the Resource Manager to release the container
+   * @param containerReleaseRequest containers to release
+   */
+  @Subscribe
+  public void handleContainerReleaseRequest(ContainerReleaseRequest containerReleaseRequest) {
+    for (Container container : containerReleaseRequest.getContainers()) {
+      LOGGER.info(String.format("Releasing container %s running on %s", container.getId(), container.getNodeId()));
+
+      // record that this container was explicitly released so that a new one is not spawned to replace it
+      this.releasedContainerSet.add(container.getId());
```

Comment:
If the ordering of inserting into the "blacklist" before firing the "releaseAssignedContainer" call matters, please document why.

Reply:
Are you always assume the handleContainerCompletion() logic would happen after releasedContainerSet.add() ? What if they happened in parallel? For example, a container encountered an issue so it triggers the handleContainerCompletion logic, which does nothing to releasedContainerSet, but suddenly the YarnAutoScaling service also thinks it needs to recycle the container, and it happened to choose the container that is in the middle of handleContainerCompletion processing? Then you will add the problematic container to the releasedContainerSet and it will never be cleaned up?

Reply:
I think we may also need to consider to update containerMap in this case? Because the containerMap was used for your container target number shrinking logic. We don't want to have a race condition here.